eclair-core/src/main/scala/fr/acinq/eclair/router/Graph.scala

```diff
@@ -25,6 +24,109 @@ object Graph {
+    val candidates = new mutable.PriorityQueue[WeightedPath]
+
+    // find the shortest path, k = 0
+    val shortestPath = dijkstraShortestPath(graph, sourceNode, targetNode, amountMsat, ignoredEdges, extraEdges)
+    shortestPaths += WeightedPath(shortestPath, pathCost(shortestPath, amountMsat))
+
+    // main loop
+    for(k <- 1 until pathsToFind) {
+
+      if ( !allSpurPathsFound ) {
+
+        // for every edge in the path
+        for (i <- shortestPaths(k - 1).path.indices) {
+
+          // select the spur node as the i-th element of the k-th previous shortest path (k -1)
+          val spurEdge = shortestPaths(k - 1).path(i)
+
+          // select the subpath from the source to the spur node of the k-th previous shortest path
+          val rootPathEdges = subList(shortestPaths(k - 1).path, 0, i).toList
+
+          // links to be removed that are part of the previous shortest path and which share the same root path
+          val edgesToIgnore = shortestPaths.flatMap { weightedPath =>
+            if (subList(weightedPath.path, 0, i) == rootPathEdges) {
+              weightedPath.path(i).desc :: Nil
+            } else {
+              Nil
+            }
+          }
+
+          // find the "spur" path, a subpath going from the spur edge to the target avoiding previously found subpaths
+          val spurPath = dijkstraShortestPath(graph, spurEdge.desc.a, targetNode, amountMsat, ignoredEdges ++ edgesToIgnore.toSet, extraEdges)
+
+          // if there wasn't a path the spur will be empty
+          if(spurPath.nonEmpty) {
+
+            // candidate k-shortest path is made of the rootPath and the new spurPath
+            val totalPath = rootPathEdges.head.desc.a == spurPath.head.desc.a match {
+              case true => rootPathEdges.tail ++ spurPath // if the heads are the same node, drop it from the rootPath
+              case false => rootPathEdges ++ spurPath
+            }
+
+            //val totalPath = concat(rootPathEdges, spurPath.toList)
+            val candidatePath = WeightedPath(totalPath, pathCost(totalPath, amountMsat))
+
+            if (!shortestPaths.contains(candidatePath) && !candidates.exists(_ == candidatePath)) {
+              candidates.enqueue(candidatePath)
+            }
+
+          }
+        }
+      }
+
+      if(candidates.isEmpty) {
+        // handles the case of having exhausted all possible spur paths and it's impossible to reach the target from the source
+        allSpurPathsFound = true
+      } else {
+        // move the best candidate to the shortestPaths container
+        shortestPaths += candidates.dequeue()
+      }
+    }
+
+    shortestPaths
+  }
+
+  // Calculates the cost of a path, direct channels with the source will have a cost of 0 (pay no fees), only the first
+  // edge in the list is a direct channel
+  def pathCost(path: Seq[GraphEdge], amountMsat: Long): Long = {
+    path.drop(1).reverse.foldLeft(amountMsat) { (cost, edge) =>
```

Comment:
do we really need the `reverse`?

Reply:
I think so, or are you suggesting to use 'foldRight' to avoid doing a reverse? (Just peeped into the scala implementations, foldRight calls reverse before applying foldLeft)

Reply:
Done with foldRight